### ISSUE TITLE
Expose malloc statistics

### ISSUE THREAD

**Issue by pitrou** [giampaolo/psutil#1275]
I don't know if it's exactly in the scope, but just in case: it could be useful to expose per-platform malloc() statistics, for example using `mallinfo()` on GNU/Linux:
http://man7.org/linux/man-pages/man3/mallinfo.3.html

**Comment by giampaolo** [comment]
Hey Antoine! I don't know either but it looks kinda too low-levelish. Do you have a use case?

**Comment by pitrou** [comment]
My use case was debugging memory use here: https://bugs.python.org/issue33444
I'm not sure there's a production use :-) Though I'm sure authors of sophisticated parallel computing frameworks (such as Dask -- @mrocklin, @jakirkham or @ogrisel) would like to be able to diagnose whether a memory consumption problem is a memory leak caused by Python objects.

**Comment by jakirkham** [comment]
Not having used `mallinfo` much, can't say much about it. That said, I can grok why it would be useful.

Just to understand the use case a bit more. Are you hoping to call `mallinfo` using `ctypes` (?) within a process that you are trying to debug? Namely by inserting `mallinfo` calls wherever it seems to matter to get some intuition about how memory usage is changing over time?

**Comment by pitrou** [comment]
What I mean is that a framework (like Dask) which already exposes memory usage statistics (such as RSS) could expose additional useful information thanks to `mallinfo`. The main info of interest IMHO is how much memory is kept in the allocator despite being nominally released (because of fragmentation).

**Comment by mrocklin** [comment]
I've been trying to track down memory leaks when using Pandas in parallel in situations that look similar to the bug report pointed to by @pitrou .  I agree that increased visibility would be of value.

**Comment by jakirkham** [comment]
To rephrase my question in the Dask context, does this make sense to call from the nanny process or does it only make sense in the worker process?

**Comment by pitrou** [comment]
Only sense in the worker process, IMO.

**Comment by giampaolo** [comment]
As for usefulness of this, I'm skeptical. We are already able to determine memory leaks by using `Process.memory_info().rss` or `Process.memory_full_info().uss` before and after a function call. Knowing detailed stats about `malloc()` specifically looks like something which is more useful to debug a C program.
Also, it seems `mallinfo()` is basically deprecated: https://stackoverflow.com/questions/40878169/64-bit-capable-alternative-to-mallinfo

**Comment by pitrou** [comment]
>  We are already able to determine memory leaks by using Process.memory_info().rss or Process.memory_full_info().uss before and after a function call. 

See https://bugs.python.org/issue33444. A higher `rss` tells you that there may be a memory leak, not that there is one. Python uses `malloc()` for all allocations > 512 bytes. When Python releases memory to the glibc (by calling `free()`), the glibc isn't always able to return memory to the system, so `rss` appears to remain high.

> Also, it seems mallinfo() is basically deprecated

I see, perhaps `malloc_info` is more worth exposing then? Though it's return values aren't documented...

**Comment by pitrou** [comment]
In Python's case `fordblks` is the better choice (it gives you the number of bytes available in the allocator but not returned to the system). `uordblks` can be misleading because Python uses its own allocator for small-sized blocks (< 256 or 512 bytes).

USS is not better than RSS for finding out memory fragmentation.

**Comment by giampaolo** [comment]
If we figure out what `malloc_info()` values stand for and we are able to provide a concrete code sample in the doc which shows how to detect memory leaks for a function call then I suppose we could have a `psutil.malloc_memory()` or something. Basically I want there to be a concrete use case for it other than merely "reporting low level memory metrics" which very few people would use. If we can extend the same concept to other platforms (basically Windows and/or OSX) that's even better, especially because [relying on RSS/USS](https://github.com/giampaolo/psutil/blob/2ffb0ecc5e41c5486056fcd00aca274d77858860/psutil/tests/test_memory_leaks.py#L175-L182) is apparently not fully reliable.

**Comment by giampaolo** [comment]
The Windows counterpart appears to be called `_heapwalk`:
https://docs.microsoft.com/en-us/cpp/c-runtime-library/reference/heapwalk?redirectedfrom=MSDN&view=vs-2019

**Comment by giampaolo** [comment]
List of useful links/info:
https://en.wikibooks.org/wiki/Linux_Applications_Debugging_Techniques/Leaks

**Comment by giampaolo** [comment]
See comment about `mallinfo()` in #1757. It still would not grant 100% stability in detecting memory leaks, despite being more precise.

**Comment by crusaderky** [comment]
I have a real life use case on dask.distributed. The package really struggles right now to tell apart genuine memory leaks and free'd memory that hasn't been returned to the OS yet. This meausure is used in heuristics for memory rebalancing and OOM safety net systems.

Using the demo workbook attached to https://github.com/dask/distributed/pull/4774 I can reliably produce such a "leak" where I allocate a bunch of large-ish numpy arrays (160 kib each) and then free them after a few seconds. After that operation, on my GUI I read:

RSS: 1244 MiB
managed memory (Python objects tracked by dask and measured with sizeof): 749 MiB
unmanaged memory (RSS - managed: memory leaks, modules, global variables, and unreleased memory): 495 MiB

full_memory_info says nothing interesting; note how uss and rss are almost the same:

pfullmem(rss=1304870912, vms=2051538944, shared=31059968, text=2240512, lib=0, data=1326317568, dirty=0, uss=1274175488, pss=1275800576, swap=0)

however, if I run this on the process:
```python
import ctypes
libc = ctypes.CDLL("libc.so.6")
libc.malloc_stats()
```
I read:
```
Total (incl. mmap):
system bytes     = 1238429696
in use bytes     =  797209696
```
which is exactly the information I need.
Thanks to the malloc_stats output I can get these numbers out:

RSS: 1244 MiB
managed: 749 MiB
not in use: (1181 - 760) = 421 MiB
unmanaged: (1244 - 749 - 421) = 74 MiB
managed+unmanaged: (749+74) = 823 MiB

When I run my rebalancing and anti-OMM algorithms, if I had this information I could consider 823 MiB instead of 1244 MiB, knowing that the rest will be reused at the next malloc.

MacOSX Big Sur has exactly the same problem. I don't know where to get the same information though.
I could not reproduce the issue on Windows.

CC @fjetter @gjoseph92

**Comment by crusaderky** [comment]
I have a real life use case on dask.distributed. The package really struggles right now to tell apart genuine memory leaks and free'd memory that hasn't been returned to the OS yet. This meausure is used in heuristics for memory rebalancing and OOM safety net systems.

Using the demo workbook attached to https://github.com/dask/distributed/pull/4774 I can reliably produce such a "leak" where I allocate a bunch of large-ish numpy arrays (160 kib each) and then free them after a few seconds. After that operation, on my GUI I read:

RSS: 1244 MiB
managed memory (Python objects tracked by dask and measured with sizeof): 749 MiB
unmanaged memory (RSS - managed: memory leaks, modules, global variables, and unreleased memory): 495 MiB

full_memory_info says nothing interesting; note how uss and rss are almost the same:

pfullmem(rss=1304870912, vms=2051538944, shared=31059968, text=2240512, lib=0, data=1326317568, dirty=0, uss=1274175488, pss=1275800576, swap=0)

however, if I run this on the process:
```python
import ctypes
libc = ctypes.CDLL("libc.so.6")
libc.malloc_stats()
```
I read:
```
Total (incl. mmap):
system bytes     = 1238429696
in use bytes     =  797209696
```
which is exactly the information I need.
Thanks to the malloc_stats output I can get these numbers out:

RSS: 1244 MiB
managed: 749 MiB
not in use: (1181 - 760) = 421 MiB
unmanaged: (1244 - 749 - 421) = 74 MiB
managed+unmanaged: (749+74) = 823 MiB

When I run my rebalancing and anti-OMM algorithms, if I had this information I could consider 823 MiB instead of 1244 MiB, knowing that the rest will be reused at the next malloc.

MacOSX Big Sur has exactly the same problem. I don't know where to get the same information though.
I could not reproduce the issue on Windows.

CC @fjetter @gjoseph92

**Comment by giampaolo** [comment]
> if I run this on the process:
> import ctypes
> libc = ctypes.CDLL("libc.so.6")
> libc.malloc_stats()
> I read:
> 
> Total (incl. mmap):
> system bytes     = 1238429696
> in use bytes     =  797209696

On my Linux system I get this:
```
Arena 0:
system bytes     =     790528
in use bytes     =     722832
Total (incl. mmap):
system bytes     =     942080
in use bytes     =     874384
max mmap regions =          1
max mmap bytes   =     151552
0
```

Which one of these values should we expose, in your opinion? Which one is useful to detect a memory leak?

**Comment by giampaolo** [comment]
For the record, there's an old experimental branch where I exposed `mallinfo` on Linux and `_heapwalk` on Windows:
https://github.com/giampaolo/psutil/compare/malloc-info?expand=1

**Comment by giampaolo** [comment]
Looking at `malloc_stats`: it's only able to stream the output to stderr, which is a major pita. It seems the evolution of `malloc_stats` is called `malloc_info()` (https://man7.org/linux/man-pages/man3/malloc_info.3.html). Quote:

> malloc_stats function exports an XML string that describes the current state of the memory-allocation implementation in the caller.
> [...]
> The open_memstream(3) function can be used to send the output of malloc_info() directly into a buffer in memory, rather than to a file. The malloc_info() function is designed to address deficiencies in malloc_stats(3) and mallinfo(3).

The data returned, though, is undocumented, completely different than `malloc_stats`, and I'm not sure what to make of it, see:

* https://stackoverflow.com/questions/3903807/how-does-malloc-info-work
* https://stackoverflow.com/questions/19183111/how-to-interpret-malloc-info

In summary we have:

* `mallinfo`: it apparently returns the most useful info in a nice struct, but it's completely unsuitable for 64-bit machines
* `malloc_stats`: it returns info similar to `mallinfo` but it's streamed over stderr, making it unsuitable for a library, as we would have to mess with stderr without user consent
* `malloc_info`: it seems it's just broken / not useful (see above)

These must be the worst designed APIs in Linux.

**Comment by crusaderky** [comment]
> Which one of these values should we expose, in your opinion? Which one is useful to detect a memory leak?

The difference between these two is the amount of memory that will get released if you invoke malloc_trim(), or it will likely be reused if you invoke malloc:
```
Total (incl. mmap):
system bytes     =     942080
in use bytes     =     874384
```
I'm unsure what "system bytes" means. Note that it's slightly lower than USS.

**Comment by crusaderky** [comment]
@giampaolo glibc 2.33 (released 2021-02-01) adds mallinfo2, which replaces mallinfo and supports >2GiB
https://man7.org/linux/man-pages/man3/mallinfo.3.html
Note that ubuntu 20.04 currently ships glibc 2.31.

**Comment by crusaderky** [comment]
Notes:
1. this comment in the man page of mallinfo (linked above) seems incorrect to me. If I read the source code, it's clearly doing the sum of all arenas e.g. it reports the same as the "Total (incl. mmap)" paragraph of malloc_stats.
```
       Information is returned for only the main memory allocation area.
       Allocations in other arenas are excluded.  See malloc_stats(3)
       and malloc_info(3) for alternatives that include information
       about other arenas.
```
2. also by reading the git tip of the glibc source code, I spotted that the total paragraph of malloc_stats will experience an integer overflow when you get beyond 4 GiB -_-

**Comment by giampaolo** [comment]
> @giampaolo glibc 2.33 (released 2021-02-01) adds mallinfo2, which replaces mallinfo and supports >2GiB
> https://man7.org/linux/man-pages/man3/mallinfo.3.html

Sweet! I missed that. It's so recent we can't assume we can rely on it though.  When `mallinfo2` is not available we should use `malloc_info` in order to mimic the same results. I'm not sure if that's possible but `malloc_info` source code is here:
https://github.com/bminor/glibc/blob/master/malloc/malloc.c
Also, another interesting link:
https://bitbucket.org/einsteintoolkit/tickets/issues/2352/support-64-bit-numbers-for

**Comment by giampaolo** [comment]
> [EDIT]
> actually, all measures from malloc_stats will break when you go beyond 4 GiB, not just the total.
> Bug reports:
> https://sourceware.org/bugzilla/show_bug.cgi?id=27928
> https://sourceware.org/bugzilla/show_bug.cgi?id=21556

Ouch! I just saw your edit. It seems all APIs are bugged one way or another. :-\

**Comment by crusaderky** [comment]
To my knowledge malloc_info isn't buggy...

**Comment by giampaolo** [comment]
Hi all,

After many years (7) I have finally decided to give this a try. It was a lot of work, but I'm quite happy with the final result: https://github.com/giampaolo/psutil/pull/2692. We now have 2 new APIs: `psutil.heap_info()` and `psutil.heap_trim()`.

## heap_info()

```python
>>> import psutil
>>> psutil.heap_info()
pheap(heap_used=5177792, mmap_used=819200)
```

On Linux this exposes `uordblks` (heap allocations) and `hblkhd` (mmap-backed blocks) from `mallinfo2` glibc call. 

On Windows we have a third field, `heap_count`:

```python
>>> psutil.heap_info()
pheap(heap_used=5177792, mmap_used=819200, heap_count=2)
```

From the doc:

* ``heap_used``: total number of bytes currently allocated via ``malloc()`` (small allocations).
* ``mmap_used``: total number of bytes currently allocated via ``mmap()`` or via large ``malloc()`` allocations.
* ``heap_count``: (Windows only) number of private heaps created via ``HeapCreate()``.

Reference for what contributes to each field:

| Platform       | Allocation type                                                                          | Field        |
|----------------|------------------------------------------------------------------------------------------|--------------|
| UNIX / Windows | small `malloc()` ≤128 KB without `free()`                                                | `heap_used`  |
| UNIX / Windows | large `malloc()` >128 KB without `free()`, or `mmap()` without `munmap()` (UNIX)         | `mmap_used`  |
| Windows        | `HeapAlloc()` without `HeapFree()`                                                       | `heap_used`  |
| Windows        | `VirtualAlloc()` without `VirtualFree()`                                                 | `mmap_used`  |
| Windows        | `HeapCreate()` without `HeapDestroy()`                                                   | `heap_count` |

## heap_trim()

```python
>>> import psutil
>>> psutil.heap_trim()
```

`heap_trim()` requests that the underlying allocator release any unused heap pages back to the OS. The idea behind this function is to call it **before** taking measurements, as it helps normalize allocator state, reducing noise so changes in `heap_used` or `mmap_used` are caused by the tested code, not by allocator bookkeeping or fragmentation.

## Real-world use: finding a C extension leak

The workflow is simple:

1. Take a baseline snapshot of the heap.
1. Call the C extension hundreds or thousands of times.
1. Take another snapshot.
1. Compare.

```python
import psutil

psutil.heap_trim()  # reduce noise

before = psutil.heap_info()
for _ in range(400):
    my_cext_function()
after = psutil.heap_info()

print("delta heap_used =", after.heap_used - before.heap_used)
print("delta mmap_used =", after.mmap_used - before.mmap_used)
```

If `heap_used` or `mmap_used`  values increase consistently, you've found a native leak.

To reduce false positives, repeat the test multiple times, doubling the number of calls on each retry. This approach helps distinguish real leaks from random noise or transient allocations. That’s exactly what the [MemoryLeakTestCase](https://github.com/giampaolo/psutil/blob/d40164f1/psutil/test/memleak.py) class in psutil does: it runs the target function repeatedly, trims the allocator before each run, and tracks differences across retries. Memory that continues to grow after several retries is flagged as a leak.
